### PR TITLE
added move and remove for photo cluster at max zoom level

### PR DESCRIPTION
**Pull request by wronny**
as discussed in https://github.com/nextcloud/maps/issues/346 I added "Move all" and "Remove all geo data" to cluster context menu at max zoom level.
Signed-off-by: wronny <forum@wron.de>

### CODE REVIEW

**Comment by tacruc** [comment]
The code looks good to me, but I haven't tested. I'm not sure if I will find time before the weekend.

**Comment by wronny** [comment]
> I'm not sure if I will find time before the weekend.

No worries, no hurry

**Comment by tacruc** [comment]
Ok tested it, works nicely. Thanks.
There is a minor thought, I'm not shure how to solve it yet.
The left click button has two behaviors:

- more than x pictures zoom, 
- less than x pictures slideshow

The reight click botton has two different menus:
- not max zoom level
- max zoomlevel

I think we should synchronize them:

1. make both depending on the number of pictures
2. make both depinding on the zoom level 
3. make the right click menu allways the same and keep the left click action as it is.

Any thoughts on this @eneiluj @jancborchardt @wronny ?

**Comment by wronny** [comment]
thought from my point of view are:

- currently "move" as well as "remove geo data" was only available for single photos therefore I added "move all" and "remove all geo data" for the minimum photo cluster count as possible which I was only able to identify by zoom level (at max zoom level). However this guarantee that this can only being used for photos with the same geo data.

- due to the fact that other (outer) zoom level will combine photos from (many) different locations it don't make much sense to allow moving them all to one geo location. And in case of using it by mistake "move all" or "remove all geo data" would be a nightmare if using it form a lower zoom level if hundreds or more photos might be affected 

- I'm not sure if left and right click needs to be synchronized because from my point of view left click is some kind of fast and default behaviour and right click provide some kind of special or detailed options.

**Comment by tacruc** [comment]
Well, I know the details of how things are programmed and I manged to get confused by the current situation. I don't want to know, if I would be able to explain this behavior to my Grandparents. 

And at least a temporally redo action would be nice anyway, or?

**Comment by wronny** [comment]
after a long time being offline I would like to come back to this merge request. The intend of this mr was to fill the following gap: 
In case of having more than one photo at the same location (either because of same geo data or after placing photos on the map using a folder) there is no option available to move those batches of photos or to remove geo data for those batches of photos because the only available right-click menu items for batches of photos are "view" and "zoom in" - even at max zoom level. This has been discussed/complained in #346 as well as #416 . Therefore I added the new right-click menu items "move all" and "remove all geo data" to batches of photos at max zoom level only:
grafik
I limited this to max zoom level to have the same the behaviour as for other photos where "move" and "delete geo data" is only available for single photos. By limiting these menu items to max zoom level these menu items will be limited to the minimum size of photo batches. This also prevents from any usage by mistake at lower zoom levels.
@tacruc , @eneiluj maybe you can review and share your opinion?
thanks

**Comment by aleksandrmetik** [comment]
Hello @eneiluj 
Could you please help us with the review? This feature is really important for us.
Many thanks!

File: src/photosController.js [hunk 1]
```diff
@@ -358,7 +379,8 @@ PhotosController.prototype = {
     photoClusterContextmenu: function(a) {
         var layerId = a.layer._leaflet_id;
         a.layer.unbindPopup();
-        var popupContent = this._map.photosController.getPhotoClusterContextmenuPopupContent(layerId);
+        var availZoomLevels =  this._map.getMaxZoom() - this._map.getZoom();
```
Comment (tacruc) [inline src/photosController.js:382]: `var groupEditEnabled = a.layer.getChildCount() <= 20 || this._map.getZoom() === this._map.getMaxZoom()`

File: src/photosController.js [hunk 2]
```diff
@@ -367,21 +389,39 @@ PhotosController.prototype = {
         a.layer.openPopup();
     },
 
-    getPhotoClusterContextmenuPopupContent: function(layerId) {
+    getPhotoClusterContextmenuPopupContent: function (layerId, availZoomLevels) {
```
Comment (tacruc) [inline src/photosController.js:392]: `getPhotoClusterContextmenuPopupContent: function (layerId, availZoomLevels,  groupEditEnabled) {`

File: src/photosController.js [hunk 3]
```diff
@@ -367,21 +389,39 @@ PhotosController.prototype = {
...
-            '   </li>' +
+            '   </li>';
+        if (availZoomLevels > 0){
+            res +=
             '   <li>' +
             '       <button class="icon-search zoomin-photo-cluster-right-click">' +
             '           <span>' + zoomText + '</span>' +
             '       </button>' +
+            '   </li>';
+        }else{
```
Comment (tacruc) [inline src/photosController.js:411]: `if (groupEditEnabled) {`

**Review by tacruc — COMMENTED** [review]
This is how I would synchronize the behaviour.

Still keep in mind #510, which might be adopted, too.

**Comment by kds69** [comment]
Can someone please review, this is very annoying missing feature and fix would be great.

**Comment by tacruc** [comment]
Please, don't as it would there is #510 nearly finished.

**Pull request closed without merging**